src/game/client/ui.cpp

```diff
@@ -85,19 +94,51 @@ float CUI::PixelSize()
 	return Screen()->w/Graphics()->ScreenWidth();
 }
 
-void CUI::ClipEnable(const CUIRect *r)
+void CUI::ClipEnable(const CUIRect *pRect)
 {
-	m_ClipRect = *r;
-	m_Clipped = true;
-	float XScale = Graphics()->ScreenWidth()/Screen()->w;
-	float YScale = Graphics()->ScreenHeight()/Screen()->h;
-	Graphics()->ClipEnable((int)(r->x*XScale), (int)(r->y*YScale), (int)(r->w*XScale), (int)(r->h*YScale));
+	if(m_pClip == 0)
+		m_pClip = new CClip(pRect, 0);
```

Comment:
Memory handling could be improved, so it doesn't allocate memory each time. Like using a CRingBuffer.

Reply:
Or even simpler, since this is essentially a stack, an array would work just fine:
```cpp
CClip aStack[64];
int m_StackSize;
```

Reply:
Yeah, a CRingBuffer may be to much overhead for this. If think even a default of 16 clip regions should be enough for the start.